rust_snuba/src/processors/spans.rs

```diff
@@ -0,0 +1,222 @@
+use rust_arroyo::backends::kafka::types::KafkaPayload;
+use rust_arroyo::processing::strategies::InvalidMessage;
+use serde::{Deserialize, Deserializer};
+use serde_json::json;
+use serde_json::Value;
+use std::collections::BTreeMap;
+use uuid::Uuid;
+
+#[derive(Debug, Deserialize)]
+struct SpanMessage {
+    #[serde(default, deserialize_with = "deserialize_optional_uuid")]
+    event_id: Option<String>,
+    #[serde(default)]
+    organization_id: Option<u64>,
+    project_id: u64,
+    #[serde(deserialize_with = "deserialize_uuid")]
+    trace_id: String,
+    span_id: String,
+    #[serde(default)]
+    parent_span_id: Option<String>,
+    #[serde(default)]
+    segment_id: Option<String>,
+    #[serde(default)]
+    group_raw: Option<String>,
+    is_segment: bool,
+    start_timestamp_ms: u64,
+    duration_ms: u64,
+    exclusive_time_ms: u64,
+    retention_days: Option<u32>,
+    tags: BTreeMap<String, Value>,
+    sentry_tags: SentryExtractedTags,
+    #[serde(default)]
+    description: Option<String>,
+    #[serde(flatten)]
+    extra: BTreeMap<String, Value>,
+}
+
+#[derive(Debug, Deserialize)]
+struct SentryExtractedTags {
+    #[serde(default, rename(serialize = "http.method"))]
+    http_method: Option<String>,
+    #[serde(default)]
+    action: Option<String>,
+    #[serde(default)]
+    domain: Option<String>,
+    #[serde(default)]
+    module: Option<String>,
+    #[serde(default)]
+    group: Option<String>,
+    #[serde(default)]
+    system: Option<String>,
+    #[serde(default)]
+    status: Option<String>,
+    #[serde(default)]
+    status_code: Option<i32>,
+    #[serde(default)]
+    transaction: Option<String>,
+    #[serde(default, rename(serialize = "transaction.op"))]
+    transaction_op: Option<String>,
+    #[serde(default)]
+    op: Option<String>,
+    #[serde(default, rename(serialize = "transaction.method"))]
+    transaction_method: Option<String>,
+    #[serde(flatten)]
+    extra: BTreeMap<String, Value>,
+}
+
+fn deserialize_optional_uuid<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: Option<String> = Option::deserialize(deserializer)?;
+    if let Some(s) = s {
+        return Ok(Some(
+            Uuid::parse_str(&s)
+                .map_err(serde::de::Error::custom)?
+                .to_string(),
+        ));
+    }
+    Ok(None)
+}
+
+fn deserialize_uuid<'de, D>(deserializer: D) -> Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let buf = String::deserialize(deserializer)?;
+    Ok(Uuid::parse_str(&buf)
+        .map_err(serde::de::Error::custom)?
+        .to_string())
+}
+
+fn enforce_retention(_: &SpanMessage) -> Option<u32> {
+    Some(90)
+}
+
+fn process_span_meta(event: &SpanMessage, processed: &mut BTreeMap<&str, Value>) {
+    processed.insert("trace_id", json!(event.trace_id));
+    processed.insert(
+        "span_id",
+        json!(u128::from_str_radix(event.span_id.as_str(), 16).unwrap()),
+    );
```

Comment:
Maybe these need to be validated during deserialization? I think unwrap would panic when it's None, right?

Reply:
you can try but you can also just return InvalidMessage like you already do upon failed deserialization